README.md
Overview
--------

RootsExtremaInflection is a package that finds roots, extrema and
inflection points of a planar curve which is given as a data frame of
discrete (xi,yi) points.

Basic functions are:

-   `rootxi()` Taylor Regression Estimator for roots
-   `extremexi()` Taylor Regression Estimator for extrema
-   `inflexi()` Taylor Regression Estimator for inflections
-   `classify_curve()` curve classification by convexity and shape type
-   `findmaxbell()` Bell Extreme Finding Estimator for symmetric extrema
-   `findmaxtulip()` Tulip Extreme Finding Estimator for symmetric
    extrema
-   `findextreme()` Integration Extreme Finding Estimator for all types
    of extrema
-   `findroot()` Integration Root Finding Estimator for roots
-   `scan_curve()` scans a not noisy curve and finds all roots and
    extrema, inflections between them
-   `scan_noisy_curve()` scans a noisy curve and finds all roots and
    extrema, inflections between them

Installation
------------

``` r
# Install with dependencies:
install.packages('RootsExtremaInflection',dependencies=TRUE)
```

Usage
-----

``` r
library(RootsExtremaInflection)

Load data:
data(xydat)
#
#Extract x and y variables:
x=xydat$x;y=xydat$y
#
#Find root, plot results, print Taylor coefficients and rho estimation:
b<-rootxi(x,y,1,length(x),5,5,plots=TRUE);b$an;b$froot;
#
#Find extreme, plot results, print Taylor coefficients and rho estimation:
c<-extremexi(x,y,1,length(x),5,5,plots=TRUE);c$an;c$fextr;
#
#Find inflection point, plot results, print Taylor coefficients and rho estimation:
d<-inflexi(x,y,1,length(x),5,5,plots=TRUE);d$an;d$finfl;
# Create a relative big data set...
f=function(x){3*cos(x-5)};xa=0.;xb=9;
set.seed(12345);x=sort(runif(5001,xa,xb));r=0.1;y=f(x)+2*r*(runif(length(x))-0.5);
#
#Find root, plot results, print Taylor coefficients and rho estimation in parallel:
#b1<-rootxi(x,y,1,round(length(x)/2),5,5,plots=TRUE,doparallel = TRUE);b1$an;b1$froot;
# Available workers are 12 
# Time difference of 5.838743 secs
#           2.5 %       97.5 %           an
# a0 -0.006960052  0.004414505 -0.001272774
# a1 -2.982715739 -2.933308292 -2.958012016
# a2 -0.308844145 -0.213011162 -0.260927654
# a3  0.806555336  0.874000586  0.840277961
# a4 -0.180720951 -0.161344935 -0.171032943
# a5  0.007140500  0.009083859  0.008112180
# [1] 177.0000000   0.2924279
# Compare with exact root = 0.2876110196
#Find extreme, plot results, print Taylor coefficients and rho estimation in parallel:
#c1<-extremexi(x,y,1,round(length(x)/2),5,5,plots=TRUE,doparallel = TRUE);c1$an;c1$fextr;
# Available workers are 12 
# Time difference of 5.822514 secs
#            2.5 %       97.5 %           an
# a0 -3.0032740050 -2.994123850 -2.998698927
# a1 -0.0006883998  0.012218393  0.005764997
# a2  1.4745326519  1.489836668  1.482184660
# a3 -0.0340626683 -0.025094859 -0.029578763
# a4 -0.1100798736 -0.105430525 -0.107755199
# a5  0.0071405003  0.009083859  0.008112180
# [1] 1022.000000    1.852496
# Compare with exact extreme = 1.858407346
#Find inflection point, plot results, print Taylor coefficients and rho estimation in parallel:
#d1<-inflexi(x,y,1090,2785,5,5,plots=TRUE,doparallel = TRUE);d1$an;d1$finfl;
# Available workers are 12 
# Time difference of 4.343851 secs
#           2.5 %       97.5 %            an
# a0 -0.008238016  0.002091071 -0.0030734725
# a1  2.995813560  3.023198534  3.0095060468
# a2 -0.014591465  0.015326175  0.0003673549
# a3 -0.531029710 -0.484131902 -0.5075808056
# a4 -0.008253975  0.007556465 -0.0003487551
# a5  0.016126428  0.034688019  0.0254072236
# [1] 800.000000   3.427705
# Compare with exact inflection = 3.429203673
# Or execute rootexinf() and find a set of them at once and in same time:
#a<-rootexinf(x,y,100,round(length(x)/2),5,plots = TRUE,doparallel = TRUE);
#a$an0;a$an1;a$an2;a$frexinf;
# Available workers are 12 
# Time difference of 5.565372 secs
#           2.5 %      97.5 %           an0
# a0 -0.008244278  0.00836885  6.228596e-05
# a1 -2.927764078 -2.84035634 -2.884060e+00
# a2 -0.447136449 -0.30473094 -3.759337e-01
# a3  0.857290490  0.94794071  9.026156e-01
# a4 -0.198104383 -0.17360676 -1.858556e-01
# a5  0.008239609  0.01059792  9.418764e-03
#           2.5 %      97.5 %          an1
# a0 -3.005668018 -2.99623116 -3.000949590
# a1 -0.003173501  0.00991921  0.003372854
# a2  1.482600580  1.50077450  1.491687542
# a3 -0.034503271 -0.02551597 -0.030009618
# a4 -0.115396537 -0.10894117 -0.112168855
# a5  0.008239609  0.01059792  0.009418764
#           2.5 %       97.5 %          an2
# a0  0.083429390  0.092578772  0.088004081
# a1  3.007115452  3.027343849  3.017229650
# a2 -0.009867779  0.006590042 -0.001638868
# a3 -0.517993955 -0.497886933 -0.507940444
# a4 -0.043096158 -0.029788902 -0.036442530
# a5  0.008239609  0.010597918  0.009418764
#            index     value
# root          74 0.2878164
# extreme      923 1.8524956
# inflection  1803 3.4604842
#
## Next examples are for the
## Legendre polynomial of 5th order:
#
f=function(x){(63/8)*x^5-(35/4)*x^3+(15/8)*x} 
#
### findextreme()
#
## True extreme point p=0.2852315165, y=0.3466277
x=seq(0,0.7,0.001);y=f(x)
plot(x,y,pch=19,cex=0.5)
a=findextreme(x,y)
a
##        x1        x2       chi    yvalue 
## 0.2840000 0.2860000 0.2850000 0.3466274 
sol=a['chi']
abline(h=0)
abline(v=sol)
abline(v=a[1:2],lty=2)
abline(h=f(sol),lty=2)
points(sol,f(sol),pch=17,cex=2)
#
## The same function with noise from U(-0.05,0.05)
set.seed(2019-07-26);r=0.05;y=f(x)+runif(length(x),-r,r)
plot(x,y,pch=19,cex=0.5)
a=findextreme(x,y)
a
##        x1        x2       chi    yvalue 
## 0.2890000 0.2910000 0.2900000 0.3895484 
sol=a['chi']
abline(h=0)
abline(v=sol)
abline(v=a[1:2],lty=2)
abline(h=f(sol),lty=2)
points(sol,f(sol),pch=17,cex=2)
#
### findroot()
#
x=seq(0.2,0.8,0.001);y=f(x);ya=abs(y)
plot(x,y,pch=19,cex=0.5,ylim=c(min(y),max(ya)))
abline(h=0);
lines(x,ya,lwd=4,col='blue')
rt=findroot(x,y)
rt
##           x1            x2           chi        yvalue 
## 5.370000e-01  5.400000e-01  5.385000e-01 -7.442574e-05 
abline(v=rt['chi'])
abline(v=rt[1:2],lty=2);abline(h=rt['yvalue'],lty=2)
points(rt[3],rt[4],pch=17,col='blue',cex=2)
#
## Same curve but with noise from U(-0.5,0.5)
#
set.seed(2019-07-24);r=0.05;y=f(x)+runif(length(x),-r,r)
ya=abs(y)
plot(x,y,pch=19,cex=0.5,ylim=c(min(y),max(ya)))
abline(h=0)
points(x,ya,pch=19,cex=0.5,col='blue')
rt=findroot(x,y)
rt
##         x1          x2         chi      yvalue 
## 0.53400000  0.53700000  0.53550000 -0.01762159 
abline(v=rt['chi'])
abline(v=rt[1:2],lty=2);abline(h=rt['yvalue'],lty=2)
points(rt[3],rt[4],pch=17,col='blue',cex=2)
#
### scan_curve()
#
x=seq(-1,1,0.001);y=f(x)
plot(x,y,pch=19,cex=0.5)
abline(h=0)
rall=scan_curve(x,y)
rall$study
rall$roots
##          x1     x2           chi        yvalue
## [1,] -0.907 -0.905 -9.060000e-01  1.234476e-03
## [2,] -0.540 -0.537 -5.385000e-01  7.447856e-05
## [3,] -0.001  0.001  5.551115e-17  1.040844e-16
## [4,]  0.537  0.540  5.385000e-01 -7.444324e-05
## [5,]  0.905  0.907  9.060000e-01 -1.234476e-03
rall$extremes
##          x1     x2    chi     yvalue
## [1,] -0.766 -0.764 -0.765  0.4196969
## [2,] -0.286 -0.284 -0.285 -0.3466274
## [3,]  0.284  0.286  0.285  0.3466274
## [4,]  0.764  0.766  0.765 -0.4196969
rall$inflections
##          x1     x2           chi        yvalue
## [1,] -0.579 -0.576 -5.775000e-01  9.659939e-02
## [2,] -0.001  0.001  5.551115e-17  1.040829e-16
## [3,]  0.576  0.579  5.775000e-01 -9.659935e-02
#
### scan_noisy_curve()
#
x=seq(-1,1,0.001)
set.seed(2019-07-26);r=0.05;y=f(x)+runif(length(x),-r,r)
plot(x,y,pch=19,cex=0.5)
rn=scan_noisy_curve(x,y)
rn
## $study
##       j  dj interval   i1   i2  root
## 3    97 351     TRUE   97  448 FALSE
## 18  477 502     TRUE  477  979 FALSE
## 39 1021 505     TRUE 1021 1526 FALSE
## 54 1558 343     TRUE 1558 1901 FALSE
## 
## $roots_average
##       x1     x2     chi       yvalue
## 1 -0.906 -0.904 -0.9050 -0.002342389
## 2 -0.553 -0.524 -0.5385  0.005003069
## 3 -0.022  0.020 -0.0010  0.003260937
## 4  0.525  0.557  0.5410 -0.007956680
## 5  0.900  0.911  0.9055 -0.008015683
## 
## $roots_optim
##       x1     x2     chi       yvalue
## 1 -0.909 -0.901 -0.9050 -0.023334404
## 2 -0.531 -0.527 -0.5290  0.029256059
## 3  0.001  0.003  0.0020  0.001990572
## 4  0.530  0.565  0.5475  0.019616283
## 5  0.909  0.912  0.9105  0.009288338
## 
## $extremes
##          x1     x2     chi     yvalue
## [1,] -0.773 -0.766 -0.7695  0.4102010
## [2,] -0.280 -0.274 -0.2770 -0.3804006
## [3,]  0.308  0.316  0.3120  0.3372764
## [4,]  0.741  0.744  0.7425 -0.4414494
## 
## $inflections
##          x1     x2     chi       yvalue
## [1,] -0.772 -0.275 -0.5235 -0.076483193
## [2,] -0.275  0.281  0.0030 -0.007558037
## [3,]  0.301  0.776  0.5385  0.018958334
#

```

Why should I use RootsExtremaInflection package in R?
-----------------------------------------------------

-   Because it can give you a reliable estimation by using Taylor
    Regression, if you prefer non parametric statistical methods
-   Because it can find the desired point(s) by using geometric and
    analytic methods, without any kind of model adoption
-   Because it is the core of a new field called ‘Noisy Numerical
    Analysis’ which combines Geometry, Calculus, Numerical Analysis and
    Statistics in order to solve known problems of Numerical Analysis

Contact
-------

Please send comments, suggestions or bug breports to
dchristop$econ.uoa.gr


NEWS.md
What’s new in version 1.2.1
---------------------------

### New functions added for root and extrema estimation

-   `classify_curve()` classifies a curve by convexity and shape type
-   `findmaxbell()` finds maximum for a bell curve
-   `findmaxtulip()` finds maximum for a tulip curve
-   `symextreme()` finds maximum for a symmetric curve
-   `findextreme()` finds extreme point for a general curve
-   `findroot()` finds root of a curve
-   `scan_curve()` scans a not noisy curve and finds all roots and
    extrema, inflections between them
-   `scan_noisy_curve()` scans a noisy curve and finds all roots and
    extrema, inflections between them

It is the implementation of Bell Extreme Finding Estimator (BEFE),
Integration Extreme Finding Estimator (IEFE) and Integration Root
Finding Estimator (IRFE) for roots, extrema and inflections of a curve,
see:

-   Demetris T. Christopoulos (2019). New methods for computing extremes
    and roots of a planar curve: introducing Noisy Numerical Analysis
    (2019). ResearchGate.
    <a href="http://dx.doi.org/10.13140/RG.2.2.17158.32324" class="uri">http://dx.doi.org/10.13140/RG.2.2.17158.32324</a>


R/scan_noisy_curve.R
scan_noisy_curve=function(x,y,noise=NULL,rootsoptim=TRUE,findextremes=TRUE,findinflections=TRUE,silent=FALSE,plots=TRUE){
  # Function for computing roots, extrema and inflections for a noisy curve
  # Useful functions
  round2 <- function(x) { trunc(x + sign(x) * 0.5)}
  # Check if curve is without error
  check_noise=function(y,nlag){
    sds=sapply(1:nlag, function(i,y){round(sd(diff(y,i)),6)},y=y)
    if(sds[1]==0){
      return(FALSE)
    }else{
      sumsds=sum(round(sds/sds[1],1))
      nseq=nlag*(nlag+1)/2
      ifelse(abs(sumsds-nseq)>(nlag-2),{return(TRUE)},{return(FALSE)})
    }
  }
  # Interpolation
  polint2=function(x,x1,x2,x3,y1,y2,y3){
    y1*(x-x2)*(x-x3)/((x1-x2)*(x1-x3))+y2*(x-x1)*(x-x3)/((x2-x1)*(x2-x3))+y3*(x-x1)*(x-x2)/((x3-x1)*(x3-x2))
  }
  # Check for noise if not given true
  if(is.null(noise)){
    ifelse(length(x)>10,{nlag=5},{nlag=3})
    noise=check_noise(y,nlag)
  }
  #
  if(noise){
    scan1=scan_curve(x,y,findroots = FALSE,findextremes = FALSE,findinflections = FALSE,silent=TRUE,plots=plots)
    sc=scan1$study
    sc
    if(length(sc)==1){
      return(scan1)
    }else{
      if(plots){points(sc$x,sc$y,col='gold',pch=19)}
      jj=as.integer(rownames(sc))      
      dj=as.integer(diff(jj))
      djlogs=round(log10(dj))
      djlogsu=unique(djlogs)
      if(sum(djlogsu)==1){one=TRUE}else{one=FALSE}
      #
      ###
      if(length(dj)!=0){
        #
        dd=data.frame("j"=jj[1:(length(jj)-1)],"dj"=dj)
        dd=dd[order(dd$dj,decreasing = T),]
        dd2=dd[dd$dj>1,]        
        if(dim(dd2)[1]>=5){
          max2=max(dd2$dj);max2
          dj2=(dd2$dj-max2)^2;dj2          
          nint=ede(1:length(dj2),dj2,0)[2]-1
        }else{
          nint=1
        }
        dint=dd[1:nint,]
        dint
        if(nint==1 | one){
          onlyone=TRUE
          jint=dint$j
          djint=dint$dj
          dr=data.frame("j"=jint,"dj"=djint)
          dr[,"interval"]=FALSE
          dr[,"i1"]=dr[,"j"]
          dr[,"i2"]=dr[,"j"]+dr[,"dj"]
          dr$root=rep(TRUE,dim(dr)[1])
          dr
          sol1=x[jint];yval1=y[jint];c(sol1,yval1)
          if(dim(dr)[1]==1){
            sol2=mean(x[dr$i1:dr$i2]);yval2=mean(y[dr$i1:dr$i2]);c(sol2,yval2)
            sol3=x[jint+djint];yval3=y[jint+djint];c(sol3,yval3)
          }else{
            sol2=mean(x[jj]);yval2=mean(y[jj]);c(sol2,yval2)
            sol3=mean(x[jint+djint]);yval3=mean(y[jint+djint]);c(sol3,yval3)
          }          
          vlogs=abs(round2(log10(abs(c(sol1,sol2,sol3)))));vlogs
          if(length(unique(vlogs))==1 | one){
            sol=mean(x[jj]);yval=mean(y[jj]);c(sol,yval)
            xyroots=data.frame("x1"=x[jj[1]],"x2"=x[jj[length(jj)]],"chi"=sol,"yvalue"=yval)
          }else{
            xyroots=data.frame("x1"=x[jj[1]],"x2"=x[jj[length(jj)]],"chi"=c(sol1,sol2,sol3),"yvalue"=c(yval1,yval2,yval3))
            xyroots=xyroots[abs(xyroots$yvalue)<1,]
          }
          xyroots
          #
          if(plots){
            abline(v=xyroots$chi,lty=2)
            points(xyroots$chi,xyroots$yvalue,col='green',pch=20,cex=2)
          }
        }else{
          onlyone=FALSE
          dd=dd[order(dd$j),]
          dd$interval=rep(FALSE,dim(dd)[1])
          dd[rownames(dd)%in%rownames(dint),"interval"]=TRUE
          dd$i1=rep(NA,dim(dd)[1])
          dd$i2=rep(NA,dim(dd)[1])
          dd[dd$interval,"i1"]=dd[dd$interval,"j"]
          dd[dd$interval,"i2"]=dd[dd$interval,"j"]+dd[dd$interval,"dj"]
          dd$root=rep(FALSE,dim(dd)[1])
          dd[!dd$interval,"root"]=TRUE
          dd
          dr=dd[dd$interval,]
          dr
          # Find roots, attempt 1
          if(dim(dr)[1]>=1 & !onlyone){
            # List of "roots" around a real root...
            #######################################
            lroots=list()
            lroots[[1]]=dd$j[1]:dr[1,"i1"]
            if(dim(dr)[1]>1){
              for(k in 2:dim(dr)[1]){
                lroots[[k]]=dr[k-1,"i2"]:dr[k,"i1"]
              }
            }
            lroots[[length(lroots)+1]]=dr[dim(dr)[1],"i2"]:dd[dim(dd)[1],'j']
            # Find average roots:
            xyroots=data.frame(t(sapply(lroots, function(lr,x,y){
              out=c(x[lr[1]],x[lr[length(lr)]],mean(x[lr]),mean(y[lr]))
              names(out)=c("x1","x2","chi","yvalue")
              return(out)
            },x=x,y=y)))
            #
            if(plots){points(xyroots$chi,xyroots$yvalue,col='green',pch=20,cex=2)}
            #
          }else{
            if(!silent){message(paste0('It seems that no roots exist at [',x[1],' , ',x[length(x)],']'))}
            xyroots=NA
          }
          #ok roots 1st attempt done.
        }
        #
        if(rootsoptim & !onlyone){
          # roots 2nd attempt now...use optimization
          # 1st root
          irangefirst=round(mean(1:dr[1,"i1"])):round(mean(dr[1,"i1"]:dr[1,"i2"]))
          x1first=x[irangefirst];y1first=y[irangefirst]
          ifelse(length(irangefirst)>=7,{rfirst=findroot(x1first,y1first)},
                 {rfirst=c("x1"=x[irangefirst[1]],x2=x[irangefirst[2]],"chi"=NA,"yvalue"=NA) })
          # indermediate roots
          rootsinter=t(sapply(1:(dim(dr)[1]-1), function(i,dextrs){
            irange=round(mean(dr[i,"i1"]:dr[i,"i2"])):round(mean(dr[i+1,"i1"]:dr[i+1,"i2"]))
            if(length(irange)>=7){
              x1=x[irange];y1=y[irange]
              cc=check_curve(x1,y1)
              cc
              out=findroot(x1,y1)
              return(out)
            }else{
              out=c("x1"=x[irange[1]],x2=x[irange[2]],"chi"=NA,"yvalue"=NA)
              return(out)
            }
          },dr))
          rootsinter
          # last root
          irangelast=round(mean(dr[dim(dr)[1],"i1"]:dr[dim(dr)[1],"i2"])):round(mean(dr[dim(dr)[1],"i2"]:length(x)))
          x1last=x[irangelast];y1last=y[irangelast]
          ifelse(length(irangelast)>=7,{rlast=findroot(x1last,y1last)},
                 {rlast=c("x1"=x[irangelast[1]],x2=x[irangelast[2]],"chi"=NA,"yvalue"=NA) })
          #
          xyroots2=as.data.frame(rbind(rfirst,rootsinter,rlast))
          rownames(xyroots2)=1:dim(xyroots2)[1]
          colnames(xyroots2)=c("x1","x2","chi","yvalue")
          xyroots2
          #
          if(plots){points(xyroots2$chi,xyroots2$yvalue,col='green',pch="|",cex=2)}
        }else if(rootsoptim & onlyone){
          xyroots2=findroot(x,y)
        }
        #
        if(findextremes){
            # Find extreme points
            if(dim(dr)[1]>=1 &!onlyone){
              xt=t(sapply(1:dim(dr)[1], function(i,dr){
                irange=dr[i,"i1"]:dr[i,"i2"];irange
                if(length(irange)>=7){
                  out=findextreme(x[irange],y[irange])
                  return(out)
                }else{
                  out=c("x1"=x[irange[1]],x2=x[irange[2]],"chi"=NA,"yvalue"=NA)
                  return(out)
                }
              },dr))
              #
              if(plots & sum(sapply(xt[1,],is.na))==0){
                points(xt[,"chi"],xt[,"yvalue"],col='black',pch=15,cex=2)
                xtx=xt[,'chi'];xty=xt[,'yvalue']
                segments(x0=xtx,y0=rep(0,length(xtx)),x1=xtx,y1=xty,lty=2,col='black')
              }
              #
            }else{
              if(!silent){message(paste0('It seems that no extreme between roots exist at [',x[1],' , ',x[length(x)],']'))}
              xt=NA
            }
        }else{
          xt=NA
        }
        # ok extremes found.
        #
        if(findinflections){  
            # Find inflection points
            if(dim(dr)[1]>1 &!onlyone){
              pinf=t(sapply(1:(dim(dr)[1]-1), function(i,dextrs){
                irange=dr[i,"i1"]:dr[i+1,"i2"]
                if(length(irange)>=4){
                  x1=x[irange];y1=y[irange]
                  cc=check_curve(x1,y1)
                  cc
                  xd=ede(x1,y1,cc$index)
                  solint=x1[xd[1:2]]
                  sol=xd[3]
                  j1=xd[1,1]
                  j3=xd[1,2]
                  j2=round(mean(c(j1,j3)))
                  c(j1,j2,j3)
                  yval=polint2(sol,x1[j1],x1[j2],x1[j3],y1[j1],y1[j2],y1[j3]);yval
                  out=c(solint,sol,yval)
                  names(out)=c("x1","x2","chi","yvalue")
                  return(out)
                }else{
                  out=c("x1"=x[irange[1]],x2=x[irange[2]],"chi"=NA,"yvalue"=NA)
                  return(out)
                }
              },dr))
              pinf
              inflects=pinf
              #
              if(plots){
                points(inflects[,"chi"],inflects[,"yvalue"],col='black',pch=18,cex=2)
                pinfx=pinf[,'chi'];pinfy=pinf[,'yvalue']
                segments(x0=pinfx,y0=rep(0,length(pinfx)),x1=pinfx,y1=pinfy,lty=2,col='black')
                
              }
              #
            }else{
              if(!silent){message(paste0('It seems that no inflections between roots exist at [',x[1],' , ',x[length(x)],']'))}
              inflects=NA
            }
        }else{
          inflects=NA
        }
        # ok inflections found and roots 2nd attempt done.
        ###
        # Return all results:
        out=list("study"=dr,"roots_average"=xyroots,"roots_optim"=xyroots2,"extremes"=xt,"inflections"=inflects)
        #
        return(out)
      }else{
        # return(sc)
        dr=data.frame("j"=sc[1,7],"dj"=0,"i1"=1,"i2"=sc[1,7],"root"=TRUE)
        rownames(dr)=sc[1,7]
        xyroots=data.frame("chi"=sc[1,1],"yvalue"=sc[1,2])
        xr2=findroot(x,y);xr2=matrix(xr2,ncol=4,nrow=1)
        xyroots2=as.data.frame(xr2)
        colnames(xyroots2)=c("x1","x2","chi","yvalue")
        if(plots){
          points(xyroots$chi,xyroots$yvalue,col='green',pch=20,cex=2)
          points(xyroots2$chi,xyroots2$yvalue,col='green',pch="|",cex=2)
          abline(v=c(xyroots$chi,xyroots2$chi),lty=2)
        }
        out=list("study"=dr,"roots_average"=xyroots,"roots_optim"=xyroots2,"extremes"=NA,"inflections"=NA)
        #
        return(out)
      }
    }
  }else{
    if(!silent){message("Curve seems to be not a noisy one, use of 'scan_curve()' now...")}
    out=scan_curve(x,y,findroots = TRUE,findextremes = findextremes,findinflections =findinflections,silent=silent,plots=plots)
    #
    return(out)
  }
}
